Sources/SyncDataProviders/Bookmarks/BookmarksProvider.swift

```diff
@@ -0,0 +1,232 @@
+//
+//  BookmarksProvider.swift
+//
+//  Copyright © 2023 DuckDuckGo. All rights reserved.
+//
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+//
+
+import Foundation
+import Bookmarks
+import Combine
+import CoreData
+import DDGSync
+import Persistence
+
+public final class BookmarksProvider: DataProviding {
+
+    public init(database: CoreDataDatabase, metadataStore: SyncMetadataStore, reloadBookmarksAfterSync: @escaping () -> Void) {
+        self.database = database
+        self.metadataStore = metadataStore
+        self.metadataStore.registerFeature(named: feature.name)
+        self.reloadBookmarksAfterSync = reloadBookmarksAfterSync
+        syncErrorPublisher = syncErrorSubject.eraseToAnyPublisher()
+    }
+
+    public let syncErrorPublisher: AnyPublisher<Error, Never>
+
+    // MARK: - DataProviding
+
+    public let feature: Feature = .init(name: "bookmarks")
+
+    public var lastSyncTimestamp: String? {
+        get {
+            metadataStore.timestamp(forFeatureNamed: feature.name)
+        }
+        set {
+            metadataStore.updateTimestamp(newValue, forFeatureNamed: feature.name)
+        }
+    }
+
+    public func prepareForFirstSync() throws {
+        lastSyncTimestamp = nil
+
+        var saveError: Error?
+
+        let context = database.makeContext(concurrencyType: .privateQueueConcurrencyType)
+
+        context.performAndWait {
+            let fetchRequest = BookmarkEntity.fetchRequest()
+            let bookmarks = (try? context.fetch(fetchRequest)) ?? []
+            for bookmark in bookmarks {
+                bookmark.modifiedAt = Date()
+            }
+
+            do {
+                try context.save()
+            } catch {
+                saveError = error
+            }
+        }
+
+        if let saveError {
+            throw saveError
```

Comment:
Aren't these errors ignored in the call chain?

Reply:
Those errors are propagated all the way up to `SyncQueue.handleResponse` and they’re being caught in [this catch clause](https://github.com/duckduckgo/BrowserServicesKit/blob/dominik/sync-engine/Sources/DDGSync/internal/SyncQueue.swift#L172-L176). That’s how I understand it :)

Reply:
I'm not sure that's the case - isn't this called from [here](https://github.com/duckduckgo/BrowserServicesKit/pull/355/files/c9dd8d2dc39833f828be3c7ebbf4a4b6baba9262#diff-7fead194fe4fc0c9151d18a137ec673550e65ddf3b22628bcaa1e5b107653020R200)? And not all of the callse to update account are handled.

Reply:
My bad, I confused this code with a similar one in `handleSyncResponse`. You’re right, I’ll tackle this tomorrow.

Reply:
Ok, I handled it by catching this error in SyncQueue and notifying dataProvider error handler (it ends up being handled the same way as errors thrown during sync). Let me know if you’d prefer a different approach.

Reply:
This is good :)